File: 0078-subsets/0078-subsets.java
class Solution {
    List<List<Integer>> result = new ArrayList<>();
    public void helper(int index, List<Integer> ds, int[] nums){
        // if(index>=nums.length ){
            result.add(new ArrayList<>(ds));
            // return;
        // }
       for(int i = index ; i < nums.length ; i++){
        ds.add(nums[i]);
        helper(i+1,ds,nums);
        ds.remove(ds.size()-1);
       }
    }

    public List<List<Integer>> subsets(int[] nums) {
        List<Integer> ds = new ArrayList<>();
        helper(0,ds,nums);
        return result;
    }
}

File: 0078-subsets/0078-subsets.cpp
class Solution {
public:
    vector<vector<int>> result;
    void helper(int index, vector<int> &ds, vector<int> &nums){
        // if(index >= nums.size()){
            result.push_back(ds);
            // return;
        // }
        for(int i = index ; i < nums.size() ; i++){
            ds.push_back(nums[i]);
            helper(i+1,ds,nums);
            ds.pop_back();
        }
    }
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<int> ds;
        helper(0,ds,nums);
        // vector<vector<int>> res = vector<vector<int>>(result.begin(),result.end());
        return result;
    }
};